From 66c360a03c1ffcb57486a8a9efb166b3bfdae88a Mon Sep 17 00:00:00 2001
From: Nicolas Bock <nicolas.bock@suse.com>
Date: Wed, 12 Dec 2018 12:55:31 +0000
Subject: [PATCH] SCRD-5556 Disable Spectre v2 (retpoline) mitigation

For nested astack deployments, SP4 kernels that mitigate Spectre v2 fail to
boot with a kernel panic. This change disables the mitigation.

This change also moves `diskimage.json` into a template so that we can
modify specific settings depending on the deployment scenario.

Change-Id: I53f9d94c3aec9f687b2f62492031196a76ceae5a
---
 .../files/sles12sp4/diskimage.json            | 94 ------------------
 .../image-build/tasks/build-packer-qcow2.yml  |  7 +-
 .../templates/sles12sp4/autoyast.xml.j2       | 11 +++
 .../templates/sles12sp4/diskimage.json.j2     | 98 +++++++++++++++++++
 4 files changed, 115 insertions(+), 95 deletions(-)
 delete mode 100644 ansible/roles/image-build/files/sles12sp4/diskimage.json
 create mode 100644 ansible/roles/image-build/templates/sles12sp4/diskimage.json.j2

diff --git a/ansible/roles/image-build/files/sles12sp4/diskimage.json b/ansible/roles/image-build/files/sles12sp4/diskimage.json
deleted file mode 100644
index e509712..0000000
--- a/ansible/roles/image-build/files/sles12sp4/diskimage.json
+++ /dev/null
@@ -1,94 +0,0 @@
-{
-  "variables": {
-    "image_name": "sles12sp4.qcow2",
-    "box_distro": "sles12sp4",
-    "iso_url": "file://{{ env `HOME` }}/sles12sp4/sles12sp4.iso",
-    "iso_checksum": "1927d1b8b320839a59c69025385c61105634c83761580ca699a70334937de658",
-    "iso_checksum_type": "sha256",
-    "disk_size": "4120",
-    "output_directory": "{{ env `PWD` }}/output",
-    "boxes_directory": "{{ env `PWD` }}/boxes",
-    "vagrantfile_template": "{{ env `PWD` }}/Vagrantfile",
-    "base_dir": "{{ env `PWD` }}/",
-    "username": "{{ env `ARDANAUSER` }}",
-    "password": "{{ env `ARDANAUSER` }}",
-    "home_base": "{{ env `ARDANA_USER_HOME_BASE` }}",
-    "http_proxy": "{{ env `http_proxy` }}",
-    "https_proxy": "{{ env `https_proxy` }}",
-    "no_proxy": "{{ env `no_proxy` }}",
-    "HTTP_PROXY": "{{ env `HTTP_PROXY` }}",
-    "HTTPS_PROXY": "{{ env `HTTPS_PROXY` }}",
-    "NO_PROXY": "{{ env `NO_PROXY` }}"
-  },
-
-  "builders": [
-    {
-      "name": "{{ user `image_name` }}",
-      "output_directory": "{{ user `output_directory` }}",
-      "type": "qemu",
-      "iso_url": "{{ user `iso_url` }}",
-      "iso_checksum": "{{ user `iso_checksum` }}",
-      "iso_checksum_type": "{{ user `iso_checksum_type` }}",
-      "disk_size": "{{ user `disk_size` }}",
-      "disk_cache": "unsafe",
-      "disk_discard": "unmap",
-      "disk_interface": "virtio-scsi",
-      "format": "qcow2",
-      "headless": true,
-      "accelerator": "kvm",
-      "http_directory": "{{ user `base_dir` }}/http",
-      "ssh_username": "{{ user `username` }}",
-      "ssh_password": "{{ user `password` }}",
-      "ssh_wait_timeout": "1200s",
-      "shutdown_command": "echo '/sbin/halt -h -p' > shutdown.sh; echo '{{ user `password` }}'|sudo -S sh 'shutdown.sh'",
-      "boot_command": [
-        "<esc><enter><wait>",
-        "linux netdevice=eth0 netsetup=dhcp install=cd:/<wait>",
-        " autoyast=http://{{ .HTTPIP }}:{{ .HTTPPort }}/autoyast.xml<wait>",
-        " self_update=0<wait>",
-        "<enter><wait>"
-      ],
-      "qemuargs": [
-        [ "-m", "1024M" ]
-      ]
-    }
-  ],
-
-  "provisioners": [
-    {
-      "type": "shell",
-      "environment_vars": [
-        "VAGRANT_USER={{ user `username` }}",
-        "HOME_BASE={{ user `home_base` }}",
-        "http_proxy={{ user `http_proxy` }}",
-        "https_proxy={{ user `https_proxy` }}",
-        "no_proxy={{ user `no_proxy` }}",
-        "HTTP_PROXY={{ user `HTTP_PROXY` }}",
-        "HTTPS_PROXY={{ user `HTTPS_PROXY` }}",
-        "NO_PROXY={{ user `NO_PROXY` }}"
-      ],
-      "execute_command": "echo '{{ user `password` }}'|{{ .Vars }} sudo -E -S sh '{{ .Path }}'",
-      "scripts": [
-        "{{ user `base_dir` }}/scripts/base.sh",
-        "{{ user `base_dir` }}/scripts/vagrant.sh",
-        "{{ user `base_dir` }}/scripts/ardana.sh",
-        "{{ user `base_dir` }}/scripts/cleanup.sh",
-        "{{ user `base_dir` }}/scripts/zerodisk.sh"
-      ]
-    }
-  ],
-
-  "post-processors": [
-    {
-      "type": "vagrant",
-      "compression_level": 9,
-      "output": "{{ user `boxes_directory` }}/{{ user `box_distro` }}-{{ user `username` }}_{{ .Provider }}.box",
-      "keep_input_artifact": true,
-      "override": {
-        "libvirt": {
-          "vagrantfile_template": "{{ user `vagrantfile_template` }}"
-        }
-      }
-    }
-  ]
-}
diff --git a/ansible/roles/image-build/tasks/build-packer-qcow2.yml b/ansible/roles/image-build/tasks/build-packer-qcow2.yml
index 98e6bab..223326a 100644
--- a/ansible/roles/image-build/tasks/build-packer-qcow2.yml
+++ b/ansible/roles/image-build/tasks/build-packer-qcow2.yml
@@ -91,6 +91,11 @@
   when:
     - not _sha256_checksum_stat_result.stat.exists
 
+- name: image-build | build-packer-qcow2 | Create packer template
+  template:
+    src: "{{ playbook_dir }}/roles/image-build/templates/{{ image_build_name }}/diskimage.json.j2"
+    dest: "{{ image_build_http_dir }}/diskimage.json"
+
 - name: image-build | build-packer-qcow2 | Run packer
   shell: |
     set -eu
@@ -113,7 +118,7 @@
         -var "password={{ image_build_password }}" \
         -var "disk_size={{ image_build_disk_size }}" \
         -var "base_dir={{ ( 'roles/image-build/files/' ~ image_build_name ) | realpath }}" \
-        "{{ ( 'roles/image-build/files/' ~ image_build_name ) | realpath }}/diskimage.json"
+        "{{ image_build_http_dir }}/diskimage.json"
     echo packer built
 
     qemu-img convert -c -f qcow2 -O qcow2 -o cluster_size=2M \
diff --git a/ansible/roles/image-build/templates/sles12sp4/autoyast.xml.j2 b/ansible/roles/image-build/templates/sles12sp4/autoyast.xml.j2
index 634cbc0..8393a3e 100755
--- a/ansible/roles/image-build/templates/sles12sp4/autoyast.xml.j2
+++ b/ansible/roles/image-build/templates/sles12sp4/autoyast.xml.j2
@@ -22,6 +22,17 @@
 #}
 <!DOCTYPE profile>
 <profile xmlns="http://www.suse.com/1.0/yast2ns" xmlns:config="http://www.suse.com/1.0/configns">
+  <bootloader>
+    <global>
+      <timeout config:type="integer">2</timeout>
+      <append>spectre_v2=off</append>
+    </global>
+  </bootloader>
+  <report>
+    <warnings>
+      <timeout config:type="integer">0</timeout>
+    </warnings>
+  </report>
   <deploy_image>
     <image_installation config:type="boolean">false</image_installation>
   </deploy_image>
diff --git a/ansible/roles/image-build/templates/sles12sp4/diskimage.json.j2 b/ansible/roles/image-build/templates/sles12sp4/diskimage.json.j2
new file mode 100644
index 0000000..705d286
--- /dev/null
+++ b/ansible/roles/image-build/templates/sles12sp4/diskimage.json.j2
@@ -0,0 +1,98 @@
+{
+  "variables": {
+    "image_name": "sles12sp4.qcow2",
+    "box_distro": "sles12sp4",
+    "iso_url": "file://{{ "{{" }} env `HOME` {{ "}}" }}/sles12sp4/sles12sp4.iso",
+    "iso_checksum": "1927d1b8b320839a59c69025385c61105634c83761580ca699a70334937de658",
+    "iso_checksum_type": "sha256",
+    "disk_size": "4120",
+    "output_directory": "{{ "{{" }} env `PWD` {{ "}}" }}/output",
+    "boxes_directory": "{{ "{{" }} env `PWD` {{ "}}" }}/boxes",
+    "vagrantfile_template": "{{ "{{" }} env `PWD` {{ "}}" }}/Vagrantfile",
+    "base_dir": "{{ "{{" }} env `PWD` {{ "}}" }}/",
+    "username": "{{ "{{" }} env `ARDANAUSER` {{ "}}" }}",
+    "password": "{{ "{{" }} env `ARDANAUSER` {{ "}}" }}",
+    "home_base": "{{ "{{" }} env `ARDANA_USER_HOME_BASE` {{ "}}" }}",
+    "http_proxy": "{{ "{{" }} env `http_proxy` {{ "}}" }}",
+    "https_proxy": "{{ "{{" }} env `https_proxy` {{ "}}" }}",
+    "no_proxy": "{{ "{{" }} env `no_proxy` {{ "}}" }}",
+    "HTTP_PROXY": "{{ "{{" }} env `HTTP_PROXY` {{ "}}" }}",
+    "HTTPS_PROXY": "{{ "{{" }} env `HTTPS_PROXY` {{ "}}" }}",
+    "NO_PROXY": "{{ "{{" }} env `NO_PROXY` {{ "}}" }}"
+  },
+
+  "builders": [
+    {
+      "name": "{{ "{{" }} user `image_name` {{ "}}" }}",
+      "output_directory": "{{ "{{" }} user `output_directory` {{ "}}" }}",
+      "type": "qemu",
+      "iso_url": "{{ "{{" }} user `iso_url` {{ "}}" }}",
+      "iso_checksum": "{{ "{{" }} user `iso_checksum` {{ "}}" }}",
+      "iso_checksum_type": "{{ "{{" }} user `iso_checksum_type` {{ "}}" }}",
+      "disk_size": "{{ "{{" }} user `disk_size` {{ "}}" }}",
+      "disk_cache": "unsafe",
+      "disk_discard": "unmap",
+      "disk_interface": "virtio-scsi",
+      "format": "qcow2",
+      "headless": true,
+      "accelerator": "kvm",
+      "http_directory": "{{ "{{" }} user `base_dir` {{ "}}" }}/http",
+      "ssh_username": "{{ "{{" }} user `username` {{ "}}" }}",
+      "ssh_password": "{{ "{{" }} user `password` {{ "}}" }}",
+      "ssh_timeout": "30m",
+      "shutdown_command": "echo '/sbin/halt -h -p' > shutdown.sh; echo '{{ "{{" }} user `password` {{ "}}" }}'|sudo -S sh 'shutdown.sh'",
+      "boot_wait": "20s",
+      "boot_command": [
+        "<esc><enter><wait10s>",
+        "linux netdevice=eth0 ",
+	"netsetup=dhcp ",
+	"spectre_v2=off ",
+	"install=cd:/ ",
+        "autoyast=http://{{ "{{" }} .HTTPIP {{ "}}" }}:{{ "{{" }} .HTTPPort {{ "}}" }}/autoyast.xml ",
+        "self_update=0",
+        "<enter><wait>"
+      ],
+      "qemuargs": [
+        [ "-m", "1024M" ]
+      ]
+    }
+  ],
+
+  "provisioners": [
+    {
+      "type": "shell",
+      "environment_vars": [
+        "VAGRANT_USER={{ "{{" }} user `username` {{ "}}" }}",
+        "HOME_BASE={{ "{{" }} user `home_base` {{ "}}" }}",
+        "http_proxy={{ "{{" }} user `http_proxy` {{ "}}" }}",
+        "https_proxy={{ "{{" }} user `https_proxy` {{ "}}" }}",
+        "no_proxy={{ "{{" }} user `no_proxy` {{ "}}" }}",
+        "HTTP_PROXY={{ "{{" }} user `HTTP_PROXY` {{ "}}" }}",
+        "HTTPS_PROXY={{ "{{" }} user `HTTPS_PROXY` {{ "}}" }}",
+        "NO_PROXY={{ "{{" }} user `NO_PROXY` {{ "}}" }}"
+      ],
+      "execute_command": "echo '{{ "{{" }} user `password` {{ "}}" }}'|{{ "{{" }} .Vars {{ "}}" }} sudo -E -S sh '{{ "{{" }} .Path {{ "}}" }}'",
+      "scripts": [
+        "{{ "{{" }} user `base_dir` {{ "}}" }}/scripts/base.sh",
+        "{{ "{{" }} user `base_dir` {{ "}}" }}/scripts/vagrant.sh",
+        "{{ "{{" }} user `base_dir` {{ "}}" }}/scripts/ardana.sh",
+        "{{ "{{" }} user `base_dir` {{ "}}" }}/scripts/cleanup.sh",
+        "{{ "{{" }} user `base_dir` {{ "}}" }}/scripts/zerodisk.sh"
+      ]
+    }
+  ],
+
+  "post-processors": [
+    {
+      "type": "vagrant",
+      "compression_level": 9,
+      "output": "{{ "{{" }} user `boxes_directory` {{ "}}" }}/{{ "{{" }} user `box_distro` {{ "}}" }}-{{ "{{" }} user `username` {{ "}}" }}_{{ "{{" }} .Provider {{ "}}" }}.box",
+      "keep_input_artifact": true,
+      "override": {
+        "libvirt": {
+          "vagrantfile_template": "{{ "{{" }} user `vagrantfile_template` {{ "}}" }}"
+        }
+      }
+    }
+  ]
+}
-- 
2.19.2

